build.rs

```diff
@@ -10,40 +19,69 @@ fn main() {
 
 /// Check device feature selection
 fn check_device_feature() {
-    if !cfg!(feature = "device-selected") {
-        if cfg!(feature = "direct-call-deprecated") {
-            eprintln!(
-                "The feature you selected is deprecated, because it was split up into sub-devices.
+    // Check if the device deprecated
+    if !cfg!(feature = "device-selected") && cfg!(feature = "direct-call-deprecated") {
+        eprintln!(
+"The feature you selected is deprecated, because it was split up into sub-devices.
 
 Example: The STM32F3Discovery board has a STM32F303VCT6 chip.
 You probably used to use `stm32f303` but now functionalities for the sub-device were added.
 In this case replace it with `stm32f303xc` to make your code build again.
 
 For more information, see \x1b]8;;https://github.com/stm32-rs/stm32f3xx-hal#selecting-the-right-chip\x1b\\README -> Selecting the right chip\x1b]8;;\x1b\\."
-            );
-        } else {
-            eprintln!(
-                "This crate requires you to specify your target chip as a feature.
+        );
+        std::process::exit(1);
+    }
 
-Please select one of the following (`x` denotes any character in [a-z]):
+    // get all device variants from the metadata
+    let metadata = MetadataCommand::new().exec().unwrap();
+    let mut device_variants = metadata.root_package().unwrap().features.clone();
+    device_variants.retain(|_, dependend_features| {
+        dependend_features
+            .iter()
+            .any(|feature| *feature == "device-selected")
+    });
+    let mut device_variants = device_variants.keys().cloned().collect::<Vec<String>>();
```

Comment:
One `filter_map` call is enough?

Reply:
```rust
    let features = metadata.root_package().unwrap().features.iter();
    let mut device_variants: HashSet<String> = features
        .filter_map(|(feature, dependend_features)| {
            dependend_features
                .iter()
                .any(|dependend_feature| dependend_feature == "device-selected")
                .then(|| feature.clone())
            // `bool::then` is only available for Rust >= 1.50
            // match dependend_features
            //     .iter()
            //     .any(|feature| feature == "device-selected")
            // {
            //     true => Some(feature.clone()),
            //     false => None,
            // }
        })
        .collect();
```

Reply:
Bumping MSVR is not a big deal, so I'll go for the `bool::then` method.